Framework/Geometry/src/Objects/ShapeFactory.cpp

```diff
@@ -1416,5 +1607,92 @@ void ShapeFactory::createGeometryHandler(Poco::XML::Element *pElem, std::shared_
   geomHandler->setShapeInfo(std::move(shapeInfo));
 }
 
+/**
+ * Generates a rotation Matrix applying the x rotation then y rotation, then z
+ * rotation
+ * @param xRotation The x rotation required in radians
+ * @param yRotation The y rotation required in radians
+ * @param zRotation The z rotation required in radians
+ * @returns a matrix of doubles to use as the rotation matrix
+ */
+Kernel::Matrix<double> ShapeFactory::generateMatrix(double xRotation, double yRotation, double zRotation) {
+  Kernel::Matrix<double> xMatrix = generateXRotation(xRotation);
+  Kernel::Matrix<double> yMatrix = generateYRotation(yRotation);
+  Kernel::Matrix<double> zMatrix = generateZRotation(zRotation);
+
+  return zMatrix * yMatrix * xMatrix;
+}
+
+/**
+ *Generates the x component of the rotation matrix
+ *@param xRotation The x rotation required in radians
+ *@returns a matrix of doubles to use as the x axis rotation matrix
+ */
+Kernel::Matrix<double> ShapeFactory::generateXRotation(double xRotation) {
+  const double sinX = sin(xRotation);
+  const double cosX = cos(xRotation);
+  std::vector<double> matrixList = {1, 0, 0, 0, cosX, -sinX, 0, sinX, cosX};
+  return Kernel::Matrix<double>(matrixList);
+}
+
+/**
+ * Generates the y component of the rotation matrix
+ * @param yRotation The y rotation required in radians
+ * @returns a matrix of doubles to use as the y axis rotation matrix
+ */
+Kernel::Matrix<double> ShapeFactory::generateYRotation(double yRotation) {
+  const double sinY = sin(yRotation);
+  const double cosY = cos(yRotation);
+  std::vector<double> matrixList = {cosY, 0, sinY, 0, 1, 0, -sinY, 0, cosY};
+  return Kernel::Matrix<double>(matrixList);
+}
+
+/**
+ * Generates the z component of the rotation matrix
+ * @param zRotation The z rotation required in radians
+ * @returns a matrix of doubles to use as the z axis rotation matrix
+ */
+Kernel::Matrix<double> ShapeFactory::generateZRotation(double zRotation) {
+  const double sinZ = sin(zRotation);
+  const double cosZ = cos(zRotation);
+  std::vector<double> matrixList = {cosZ, -sinZ, 0, sinZ, cosZ, 0, 0, 0, 1};
+  return Kernel::Matrix<double>(matrixList);
+}
+
+std::string ShapeFactory::addGoniometerTag(std::vector<double> rotationMatrix, std::string xml) {
+
+  // Delete previous goniometer from xml
+  std::size_t foundGonioTag = xml.find("<goniometer");
+  if (foundGonioTag != std::string::npos) {
+    std::size_t gonioTagLength = xml.find(">", foundGonioTag + 1) - foundGonioTag;
+    xml.erase(foundGonioTag, gonioTagLength);
+  }
+
+  // Put goniometer tag in correct place in xml
+  std::size_t gonioPlace;
+  std::size_t foundType = xml.find("</type>");
+  std::size_t foundSampleGeometry = xml.find("</samplegeometry");
```

Comment:
Having thought about it, I'm not so sure we should be adding the goniometer tag in situations where the CSG shape info is wrapped in a samplegeometry tag. This happens when SetSample is called with both the Environment and the Geometry parameters populated. In this situation a container geometry and a sample geometry is read in from a sample environment .xml file and just the sample part gets overridden by the value in the Geometry parameter. I think the use case here is that the user is saying that the can is only half full of powder so they load the can and sample shapes in from the .xml file but the sample shape needs updating.

I think at the moment we're only rotating the sample shape. This opens up a potential inconsistency where the sample shape is rotated but the can isn't. In reality this is unlikely to happen I think because goniometers are normally used on single crystal instruments where there isn't a container. So rather than try to handle this, it might be safer just not to insert the goniometer tag in sample shape information read in from a sample environment file

Reply:
Do I identify that the sample shape information is read in from a sample environment file by searching for the samplegeometry tag?

Reply:
Yes that's one way of doing it. If you put the extra condition inside SetSample::setSampleShape then you could also check whether the sampleEnv parameter is empty or not. Perhaps this is more simple. Something like this:
```
  if (args) {
    const auto refFrame = experiment.getInstrument()->getReferenceFrame();
    auto xml = tryCreateXMLFromArgsOnly(*args, *refFrame);
    if (!xml.empty()) {
      std::vector<double> rotationMatrix = experiment.run().getGoniometer().getR();
      if (!sampleEnv)
        xml = Geometry::ShapeFactory().addGoniometerTag(rotationMatrix, xml);
      CreateSampleShape::setSampleShape(experiment, xml);
      return;
    }
  }
```

Reply:
I think my latest commit deals with both of these issues. I think we were actually safe anyway with regards to the sampleenv as in the case I tried, when `sampleEnv` was true, `args` was false.

Reply:
I think the SNS system test that was failing a couple of days ago calls SetSample with both the Environment and Geometry parameters populated (which corresponds to sampleEnv=true and args also being true). This is a bit of a special case. Normally you'd just call SetSample with the Environment parameter populated which is what we've been doing in our conversations earlier today.
Just want to double check the new if statement - should the two conditions be combined with && ie both need to be true for you to add the goniometer tag?

Reply:
I've changed it between OR and AND a few times, but I think you're right it should be AND.